docs/examples/word_embedding/utils.py

```diff
@@ -0,0 +1,207 @@
+
+    subword_function = nlp.vocab.create_subword_function(
+        'NGramHashes', ngrams=ngrams, num_subwords=num_subwords)
+
+    # Store subword indices for all words in vocabulary
+    idx_to_subwordidxs = list(subword_function(vocab.idx_to_token))
+    subword_lookup = SubwordLookup(len(vocab))
+    for i, subwords in enumerate(idx_to_subwordidxs):
+        subword_lookup.set(i, np.array(subwords, dtype=np.int64))
+
+    def construct_batch(centers, contexts):
+        """Create a batch for Skipgram training objective."""
+        contexts_data, contexts_row, contexts_col = contexts
+
+        negatives_shape = (len(centers), num_negatives)
+        negatives = negatives_sampler(negatives_shape).astype(np.int64)
+
+        # Remove accidental hits
+        negatives_mask = (negatives != centers.expand_dims(1))
+        negatives_mask = mx.nd.stack(
+            negatives_mask, (negatives != contexts_col.expand_dims(1)))
+        negatives_mask = negatives_mask.min(axis=0)
+        negatives_mask = negatives_mask.astype(np.float32)
+
+        contexts = contexts_col
+        data, row, col = subword_lookup.skipgram(centers.asnumpy())
+        centers = mx.nd.sparse.csr_matrix(
+            (data, (row, col)),
+            shape=(len(centers), len(vocab) + num_subwords), dtype=np.float32)
+
+        return centers, contexts, negatives, negatives_mask
+
+    batchify = nlp.data.batchify.EmbeddingCenterContextBatchify(
+        batch_size=batch_size, window_size=window, cbow=False)
+    centers_contexts = batchify(data)
+    batches = centers_contexts.transform(construct_batch)
+
+    return (batches, vocab, subword_function)
+
+
+@numba_jitclass([('idx_to_subwordidxs',
+                  numba_types.List(numba_types.int_[::1])),
+                 ('offset', numba_types.int_)])
+class SubwordLookup(object):
+    """Just-in-time compiled helper class for fast subword lookup.
+
+    Parameters
+    ----------
+    num_words : int
+         Number of tokens for which to hold subword arrays.
+
+    """
+
+    def __init__(self, num_words):
+        self.idx_to_subwordidxs = [
+            np.arange(1).astype(np.int64) for _ in range(num_words)]
+        self.offset = num_words
+
+    def set(self, i, subwords):
+        """Set the subword array of the i-th token."""
+        self.idx_to_subwordidxs[i] = subwords
+
+    def skipgram(self, indices):
+        """Get a sparse COO array of words and subwords."""
+        row = []
+        col = []
+        data = []
+        for i, idx in enumerate(indices):
+            row.append(i)
+            col.append(idx)
+            data.append(1 / (1 + len(self.idx_to_subwordidxs[idx])))
+            for subword in self.idx_to_subwordidxs[idx]:
+                row.append(i)
+                col.append(subword + self.offset)
+                data.append(1 / (1 + len(self.idx_to_subwordidxs[idx])))
+        return np.array(data), np.array(row), np.array(col)
+
+
+class SkipGramNet(mx.gluon.HybridBlock):
```

Comment:
Should this go to `gluonnlp.model`?

Reply:
We can add it to `gluonnlp.model.train`. This class is specific to word2vec and fastText only.

Reply:
Let's keep it here for now and decide about moving after other embedding models are merged